transport-native-epoll/src/test/java/io/netty/channel/epoll/EpollSocketHalfClosedReproTest.java

```diff
@@ -0,0 +1,188 @@
+                            buf.release();
+                        }
+
+                        @Override
+                        public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+                            if (evt == ChannelInputShutdownEvent.INSTANCE) {
+                                clientHalfClosedLatch.countDown();
+                            } else if (evt == ChannelInputShutdownReadComplete.INSTANCE) {
+                                clientHalfClosedAllBytesRead.countDown();
+                                ctx.close();
+                            }
+                        }
+
+                        @Override
+                        public void channelReadComplete(ChannelHandlerContext ctx) {
+                            if (bytesSinceReadComplete == 0) {
+                                clientZeroDataReadCompletes.incrementAndGet();
+                            } else {
+                                bytesSinceReadComplete = 0;
+                            }
+                            clientReadCompletes.incrementAndGet();
+                            if (bytesRead == totalServerBytesWritten) {
+                                // Bounce this through the event loop to make sure it happens after we're done
+                                // with the read operation.
+                                ch.eventLoop().execute(new Runnable() {
+                                    @Override
+                                    public void run() {
+                                        clientReadAllDataLatch.countDown();
+                                    }
+                                });
+                            } else {
+                                ch.read();
+                            }
+                        }
+
+                        @Override
+                        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+                            ctx.close();
+                        }
+                    });
+                    ch.read();
+                }
+            });
+
+            serverChannel = sb.bind().sync().channel();
+            clientChannel.set(cb.connect(serverChannel.localAddress()).sync().channel());
+            clientChannel.get().read();
+
+            checkAwait(clientReadAllDataLatch);
+
+            // Now we need to trigger server half-close
+            ((DuplexChannel) serverChildChannel.get()).shutdownOutput();
+
+            checkAwait(clientHalfClosedLatch);
+            checkAwait(clientHalfClosedAllBytesRead);
+        } finally {
+            if (clientChannel.get() != null) {
+                clientChannel.get().close().sync();
+            }
+            if (serverChannel != null) {
+                serverChannel.close().sync();
+            }
+        }
+    }
+
+    private static void checkAwait(CountDownLatch latch) throws Throwable {
+        if (!latch.await(10, TimeUnit.SECONDS)) {
```

Comment:
I would we could also just call `latch.await()` and annotate the test with a timeout.